File: merge-tonic-lib/tonic/hydration/memory_vector_index.py
"""In-memory vector index for tests and fallback development flows."""

from __future__ import annotations

from dataclasses import dataclass, field
from math import sqrt

from .vector_index import VectorFilter, VectorRecord, VectorSearchResult


def _cosine_similarity(left: list[float], right: list[float]) -> float:
    dot = 0.0
    norm_left = 0.0
    norm_right = 0.0
    for i in range(max(len(left), len(right))):
        lv = left[i] if i < len(left) else 0.0
        rv = right[i] if i < len(right) else 0.0
        dot += lv * rv
        norm_left += lv * lv
        norm_right += rv * rv
    if norm_left == 0 or norm_right == 0:
        return 0.0
    return dot / (sqrt(norm_left) * sqrt(norm_right))


def _matches_filter(metadata: dict[str, object], filter: VectorFilter | None) -> bool:
    if not filter:
        return True
    return all(metadata.get(key) == value for key, value in filter.items())


@dataclass
class MemoryVectorIndex:
    collection_name: str
    backend: str = "memory"
    _records: dict[str, VectorRecord] = field(default_factory=dict)

    def upsert(self, records: list[VectorRecord]) -> None:
        for record in records:
            self._records[record.id] = record

    def similarity_search(
        self,
        query_embedding: list[float],
        limit: int,
        *,
        filter: VectorFilter | None = None,
    ) -> list[VectorSearchResult]:
        ranked = [
            VectorSearchResult(
                id=record.id,
                document=record.document,
                metadata=record.metadata,
                embedding=record.embedding,
                score=_cosine_similarity(query_embedding, record.embedding),
            )
            for record in self._records.values()
            if _matches_filter(record.metadata, filter)
        ]
        ranked.sort(key=lambda row: row.score, reverse=True)
        return ranked[: max(0, limit)]

    def get_by_ids(self, ids: list[str]) -> list[VectorSearchResult]:
        out: list[VectorSearchResult] = []
        for id_ in ids:
            record = self._records.get(id_)
            if record is None:
                continue
            out.append(
                VectorSearchResult(
                    id=record.id,
                    document=record.document,
                    metadata=record.metadata,
                    embedding=record.embedding,
                    score=1.0,
                )
            )
        return out

    def delete(self, *, ids: list[str] | None = None, filter: VectorFilter | None = None) -> None:
        if ids:
            for id_ in ids:
                self._records.pop(id_, None)
            return
        if filter:
            doomed = [id_ for id_, record in self._records.items() if _matches_filter(record.metadata, filter)]
            for id_ in doomed:
                self._records.pop(id_, None)

File: packages/tonic-core/src/hydration/memoryVectorIndex.ts
import type { VectorFilter, VectorIndex, VectorMetadata, VectorRecord, VectorSearchResult } from "./vectorIndex";

function cosineSimilarity(a: number[], b: number[]): number {
  let dot = 0;
  let normA = 0;
  let normB = 0;
  const len = Math.max(a.length, b.length);
  for (let i = 0; i < len; i++) {
    const av = a[i] ?? 0;
    const bv = b[i] ?? 0;
    dot += av * bv;
    normA += av * av;
    normB += bv * bv;
  }
  if (normA === 0 || normB === 0) {
    return 0;
  }
  return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}

function matchesFilter(metadata: VectorMetadata | undefined, filter: VectorFilter | undefined): boolean {
  if (!filter) {
    return true;
  }
  const md = metadata ?? {};
  return Object.entries(filter).every(([key, value]) => md[key] === value);
}

export class MemoryVectorIndex implements VectorIndex {
  readonly backend = "memory";
  readonly collectionName: string;
  private readonly records = new Map<string, VectorRecord>();

  constructor(collectionName: string) {
    this.collectionName = collectionName;
  }

  async upsert(records: VectorRecord[]): Promise<void> {
    for (const record of records) {
      this.records.set(record.id, record);
    }
  }

  async similaritySearch(
    queryEmbedding: number[],
    limit: number,
    filter?: VectorFilter,
  ): Promise<VectorSearchResult[]> {
    return [...this.records.values()]
      .filter((record) => matchesFilter(record.metadata, filter))
      .map((record) => ({
        id: record.id,
        document: record.document,
        metadata: record.metadata,
        embedding: record.embedding,
        score: cosineSimilarity(queryEmbedding, record.embedding),
      }))
      .sort((a, b) => b.score - a.score)
      .slice(0, Math.max(0, limit));
  }

  async getByIds(ids: string[]): Promise<VectorSearchResult[]> {
    return ids
      .map((id) => this.records.get(id))
      .filter((record): record is VectorRecord => Boolean(record))
      .map((record) => ({
        id: record.id,
        document: record.document,
        metadata: record.metadata,
        embedding: record.embedding,
        score: 1,
      }));
  }

  async delete(params: { ids?: string[]; filter?: VectorFilter }): Promise<void> {
    if (params.ids?.length) {
      for (const id of params.ids) {
        this.records.delete(id);
      }
      return;
    }
    if (params.filter) {
      for (const [id, record] of this.records.entries()) {
        if (matchesFilter(record.metadata, params.filter)) {
          this.records.delete(id);
        }
      }
    }
  }
}
